dask/sharedict.py

```diff
@@ -0,0 +1,97 @@
+from toolz import concat, unique
+from collections import Mapping
+
+
+class ShareDict(Mapping):
+    """ A MutableMapping composed of other MutableMappings
+
+    This mapping is composed of a mapping of dictionaries
+
+    Examples
+    --------
+    >>> a = {'x': 1, 'y': 2}
+    >>> b = {'x': 10, 'z': 3}
+    >>> s = ShareDict()
+    >>> s.update(a)
+    >>> s.update(b)
+
+    >>> dict(s)  # doctest: +SKIP
+    {'x': 10, 'y': 2, 'z': 3}
+
+    These dictionaries are stored within an internal dictionary of dictionaries
+
+    >>> list(s.dicts.values())  # doctest: +SKIP
+    [{'x': 1, 'y': 2}, {'x': 10, 'z': 3}]
+
+    By default these are named by their object id.  However, you can also
+    provide explicit names.
+
+    >>> s = ShareDict()
+    >>> s.update(a=a)
+    >>> s.update(b=b)
+    >>> s.dicts  # doctest: +SKIP
+    {'a': {'x': 1, 'y': 2}, 'b': {'x': 10, 'z': 3}}
+
+    Precedence among these dicts are ordered by an internal list
+
+    >>> s.order
+    ['a', 'b']
+    """
+    def __init__(self):
+        self.dicts = dict()
+        self.order = []
```

Comment:
Indeed. You would need to delete new keys before doing an insertion to ensure they get sorted to the end. But other than that, performance would be identical to only using dict on Python >3.5.